src/nvim/regexp_defs.h

```diff
@@ -41,18 +43,38 @@
 #define NFA_ENGINE          2
 
 typedef struct regengine regengine_T;
+typedef struct regprog regprog_T;
+typedef struct reg_extmatch reg_extmatch_T;
```

Comment:
Oh, right.

Ideally, circular dependencies would be resolved by splitting things up. See @ZyX-I advice in https://github.com/neovim/neovim/issues/5321#issuecomment-292217974 and https://github.com/neovim/neovim/issues/6371

tl;dr: What you have posted _works_, so don't change it. But if you are inclined, see if a new header can be introduced to avoid this fragile ordering.